Refactor Recordset app to use Reference API #523

shinyichen opened the pull request:
Addresses issue #523 (refactor recordset to use Reference API) and #510 (update recordset to use common AlertsService)
This version depends on ermrestjs changes in pull request 166 (https://github.com/informatics-isi-edu/ermrestjs/pull/166) with the addition of Location and other changes to support recordset app, which is address in ermrestjs issue 141 (https://github.com/informatics-isi-edu/ermrestjs/issues/141)

howdyjessie commented on common/utils.js:
```diff
@@ -60,16 +60,86 @@
+                hash = hash.split("@after(")[0]; // remove modifiers from uri
+            } else if (hash.indexOf("?limit=") !== -1) {
+                hash = hash.split("?limit=")[0]; // remove modifiers from uri
             }
```
The repetition of `if hash.indexOf(some_modifier) !== -1 ... hash = hash.split(some_modifier)[0]` seems like a good candidate for a currying function here, especially if there will be additional modifiers.

howdyjessie commented on recordset/recordset.html:
```diff
@@ -2,101 +2,101 @@
+        <div ng-show="pageInfo.loading" id="spinner">
+            <img src="spinner.gif" style="display: block; margin: 0 auto; width: 32px; height: 32px; opacity: 0.8;"/>
+            <div style="margin-top: 15px;">Loading...</div>
+        </div>
```
@hongsudt This div appears in almost all of our Chaise apps. We should refactor this styling out onto the main stylesheet, and perhaps this div into its own little spinner directive. Spinners aren't exactly high priority but maybe this could be a small issue.

hongsudt replied:
@jessie can you create an issue for it?

On Aug 23, 2016 5:32 PM, "Jessie Wong" <notifications@github.com> wrote:

> In recordset/recordset.html
> <https://github.com/informatics-isi-edu/chaise/pull/546#discussion_r75977167>
> :
>
> > -                            <span ng-show="pageInfo.pageLimit !== 100" class="glyphicon glyphicon-invisible"></span>
> > -                            100</a>
> > -                        </li>
> > -                        <li><a href ng-click="pageLimit(200)">
> > -                            <span ng-show="pageInfo.pageLimit === 200" class="glyphicon glyphicon-ok"></span>
> > -                            <span ng-show="pageInfo.pageLimit !== 200" class="glyphicon glyphicon-invisible"></span>
> > -                            200</a>
> > -                        </li>
> > -                    </ul>
> > -                </span>
> > -            </div>
> > +        <!-- loading spinner -->
> > +        <div ng-show="pageInfo.loading" id="spinner">
> > +            <img src="spinner.gif" style="display: block; margin: 0 auto; width: 32px; height: 32px; opacity: 0.8;"/>
> > +            <div style="margin-top: 15px;">Loading...</div>
> > +        </div>
>
> @hongsudt <https://github.com/hongsudt> This div appears in almost all of
> our Chaise apps. We should refactor this styling out onto the main
> stylesheet, and perhaps this div into its own little spinner directive.
> Spinners aren't exactly high priority but maybe this could be a small issue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/informatics-isi-edu/chaise/pull/546/files/5454f69862efc1d6251ae913eafdf0169683170f#r75977167>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZUDHgnaOlmTw0cDzYi_tVFNtV0dHg-ks5qi5DGgaJpZM4JrMbs>
> .[image: Web Bug from
> https://github.com/notifications/beacon/AAZUDICxb5inFOeVslj3akZn7E6LbZtGks5qi5DGgaJpZM4JrMbs.gif]
> {"api_version":"1.0","publisher":{"api_key":"
> 05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":
> {"external_key":"github/informatics-isi-edu/chaise","
> title":"informatics-isi-edu/chaise","subtitle":"GitHub
> repository","main_image_url":"https://cloud.githubusercontent.com/assets/
> 143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","
> avatar_image_url":"https://cloud.githubusercontent.com/
> assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png
> ","action":{"name":"Open in GitHub","url":"https://github.
> com/informatics-isi-edu/chaise"}},"updates":{"snippets":[{"
> icon":"PERSON","message":"@howdyjessie in #546: @hongsudt This div
> appears in almost all of our Chaise apps. We should refactor this styling
> out onto the main stylesheet, and perhaps this div into its own little
> spinner directive. Spinners aren't exactly high priority but maybe this
> could be a small issue."}],"action":{"name":"View Pull Request","url":"
> https://github.com/informatics-isi-edu/chaise/pull/546/files/
> 5454f69862efc1d6251ae913eafdf0169683170f#r75977167"}}}

howdyjessie replied:
@hongsudt Here it is, for you to assign as needed: https://github.com/informatics-isi-edu/chaise/issues/550.

howdyjessie commented on recordset/recordset.html:
```diff
@@ -2,101 +2,101 @@
+                    <li><a href ng-click="pageLimit(10)">
+                        <span ng-show="pageInfo.pageLimit === 10" class="glyphicon glyphicon-ok"></span>
+                        <span ng-show="pageInfo.pageLimit !== 10" class="glyphicon glyphicon-invisible"></span>
+                        10</a>
```
Seems like these spans can be simplified to something that avoids manipulating the DOM via [`ng-class`](https://docs.angularjs.org/api/ng/directive/ngClass). Something like: `<span class="glyphicon" ng-class="{'glphyicon-ok': pageInfo.pageLimit === 10, 'glyphicon-invisible': pageInfo.pageLimit !== 10}"></span>`

howdyjessie commented on recordset/recordset.js:
```diff
@@ -89,193 +82,194 @@ angular.module('recordset', ['ermrestjs', 'chaise.navbar', 'chaise.utils', 'chai
-                pageInfo.loading = false;
+            recordsetModel.columns = $rootScope.reference.columns.map(function (column, index, array) {
+                return {"name": column.name, "displayname": column.displayname};
+            });
```
Why not just `recordsetModel.columns = $rootScope.reference.columns`? Is it something like you just only need the name and displayname properties of each column?

shinyichen commented:
I fixed jessie's first comment. I had troubles making it work with ng-class, so I didn't put that in. And finally, the columns was written before Reference API. I changed it on the branch I am currently working on.

jrchudy commented on recordset/recordset.js:
```diff
@@ -284,142 +278,81 @@ angular.module('recordset', ['ermrestjs', 'chaise.navbar', 'chaise.utils', 'chai
-                pageInfo.loading = false;
-                pageInfo.previousButtonDisabled = true;
-                pageInfo.nextButtonDisabled = true;
+    var ermrestUri = UriUtils.chaiseURItoErmrestURI($window.location);
```
In my branch, `chaise-to-ermrest-uri`, I made changes to this function so that it throws errors when catalog or table/schema are not specified. You probably want to wrap this in a `try/catch`.
